code/json.sls

```diff
@@ -30,172 +30,251 @@
-	(cond
-	  ((or (string? x)
-	       (number? x)) (write x p))
-	  ((boolean? x) (display (if x "true" "false") p))
-	  ((symbol? x) (write (if (eq? x 'null) 'null (symbol->string x))
-			      p)) ;; for convenience
-	  ((null? x) (display "null" p))
-	  ((and (list? x)
-		(pair? (car x))
-		(not (pair? (caar x))))
-	   (write-ht x p))
-	  ((list? x) (write-array x p))
-	  (else (error "Invalid JSON object in json-write" x))))
+        (cond
+         ((or (string? x)
+              (number? x)) (write x p))
+         ((boolean? x) (display (if x "true" "false") p))
+         ((symbol? x)
+          (write (if (eq? x 'null) 'null (symbol->string x))
+                 p)) ;; for convenience
+         ((null? x) (display "null" p))
+         ((and (list? x)
+               (pair? (car x))
+               (not (pair? (caar x))))
+          (write-ht x p))
+         ((list? x) (write-array x p))
+         (else (error "Invalid JSON object in json-write" x))))
 
-      (lambda (x . maybe-port)
-	(write-any x (if (pair? maybe-port) (car maybe-port) (current-output-port))))))
+      (case-lambda
+        ;; 1 argument provides default behavior
+        ((obj)
+         (set! pretty? #f)
+         (write-any obj (current-output-port)))
+        ;; 2 arguments can be obj and port or msg and obj
+        ((msg-or-obj obj-or-port)
+         (case msg-or-obj
+           ;; pretty printing with 2 arguments take no port
+           ((pretty)
+            (set! pretty? #t)
+            (set! tabstop-size 2)
+            (if (port? obj-or-port)
+                (error 'json-write
+                       "expected scheme obj not port"
+                       obj-or-port)
+                (write-any
+                 obj-or-port
+                 (current-output-port))))
+           ;; 2 arguments and no pretty printing = obj and port
+           (else (if (port? obj-or-port)
+                     (begin
+                       (set! pretty? #f)
+                       (write-any msg-or-obj obj-or-port))
+                     (error 'json-write
+                            "expected port but got"
+                            obj-or-port)))))
+        ;; 3 arguments is always for pretty printing
+        ((msg obj tabstop-or-port)
+         (set! pretty? #t)
+         (case msg
+           ((pretty)
+            (cond ((port? tabstop-or-port)
+                   (write-any obj tabstop-or-port))
+                  (else (set! tabstop-size tabstop-or-port)
+                        (write-any obj (current-output-port)))))
+           (else (error 'json-write "invalid message" msg))))
+        ;; 4 arguments is always for pretty printing
+        ((msg obj tabstop port)
+         (when (not (port? port))
+           (error 'json-write "expected port but got" port))
+         (set! pretty? #t)
+         (set! tabstop-size tabstop)
+         (case msg
+           ((pretty) (write-any obj port))
+           (else (error 'json-write "invalid message" msg)))))))
```

Comment:
Well, this case lambda isn't as long as the function itself :rofl:

I don't think it is bad to have this here, but there are alternatives.

I dislike that I have to change the values back after they have been changed in order to get back to the default values.

Another thought I had was to make this function return an object.  Then in the first call you set all of the state parameters for the json-writer and save the return value to a variable.  Then the user could send objects to the writer and get the same configuration for rendering.

It would also allow for immediate discard like:
``` scheme
((get-json-writer 'pretty 4 (current-output-port)) process-config)
```

That would set the tabstop to 4.

If we did it this way, then `get-json-writer` could utilize optional keyword args, that would making handling defaults a bit clearer.

Anyway, I am fine with either approach, so let me know what you think.

Reply:
Imo using message passing here feels like overkill. Something you could look into is "thread parameters", but I think they're likely overkill too haha. 

To reduce optional arguments, maybe you could try writing it without passing the port argument around. That way a user of the function can specify it implicitly via `with-output-to-some-port`

ps. I believe `current-output-port` is an example of a thread parameter fyi

Reply:
It can be queried by `(current-output-port)` and set by `(current-output-port some-port)`

Reply:
Okay, that is very similar to a suggestion I had when I was still PRing to your fork.

I am definitely not against that.

So, you are suggesting that we ignore ports in favored of letting the user wrap it up in port redirection when needed?

Reply:
So, I removed the port control from the case lambda, and I think it is very legible now.